File: LeetCode-Cpp/FindMedianfromDataStream.cpp
class MedianFinder {
public:
    // Küçük sayıların tutulduğu max-heap
    // Varsayılan olarak std::priority_queue max-heap'tir.
    std::priority_queue<int> max_heap;

    // Büyük sayıların tutulduğu min-heap
    std::priority_queue<int, std::vector<int>, std::greater<int>> min_heap;

    MedianFinder() {
        // Kurucu metodda özel bir başlangıç işlemi gerekmez.
    }
    
    void addNum(int num) {
        // Yeni sayıyı max_heap'e ekleyerek başla (küçük yarıya).
        max_heap.push(num);

        // max_heap'in en büyük elemanı (yani en üstteki) min_heap'in en küçük elemanından (yani en üstteki) büyükse
        // veya max_heap gereğinden büyükse, elemanları dengele.
        // max_heap.top() her zaman min_heap.top()'tan küçük veya eşit olmalı.
        // Eğer değilse, max_heap'in en büyüğünü min_heap'e taşı.
        if (!max_heap.empty() && !min_heap.empty() && max_heap.top() > min_heap.top()) {
            min_heap.push(max_heap.top());
            max_heap.pop();
        }

        // Heap'lerin boyutlarını dengele:
        // max_heap, min_heap'e eşit veya bir eleman fazla olmalı.
        // Eğer max_heap çok küçükse (min_heap'ten az elemanı varsa), min_heap'ten max_heap'e taşı.
        if (max_heap.size() < min_heap.size()) {
            max_heap.push(min_heap.top());
            min_heap.pop();
        }
        // Eğer max_heap çok büyükse (min_heap'ten birden fazla elemanı varsa), max_heap'ten min_heap'e taşı.
        else if (max_heap.size() > min_heap.size() + 1) {
            min_heap.push(max_heap.top());
            max_heap.pop();
        }
    }
    
    double findMedian() {
        if (max_heap.empty()) {
            return 0.0; // Sayı eklenmediyse, medyan tanımsızdır veya 0.0 dönebiliriz.
        }

        // Toplam sayı tek ise, medyan max_heap'in en büyük elemanıdır.
        if (max_heap.size() > min_heap.size()) {
            return static_cast<double>(max_heap.top());
        } 
        // Toplam sayı çift ise, medyan iki heap'in en üstteki elemanlarının ortalamasıdır.
        else {
            return (static_cast<double>(max_heap.top()) + min_heap.top()) / 2.0;
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: LeetCode-Python/FindMedianfromDataStream.py
class MedianFinder:

    def __init__(self):
        # min_heap: sayıların büyük yarısını tutar, en küçük eleman tepededir.
        #           (normal min-heap)
        self.min_heap = [] 
        
        # max_heap: sayıların küçük yarısını tutar, en büyük eleman tepededir.
        #           Sayıları negatif olarak saklayarak min-heap'i max-heap gibi kullanırız.
        self.max_heap = [] 
        

    def addNum(self, num: int) -> None:
        # Yeni sayıyı önce max_heap'e (küçük yarıya) ekle
        # max_heap'e -num ekleyerek bir max-heap gibi çalışmasını sağlarız.
        heapq.heappush(self.max_heap, -num)

        # max_heap'in en büyük elemanı (yani -max_heap[0]) min_heap'in en küçük elemanından
        # (yani min_heap[0]) büyükse, elemanları dengele.
        # Bu, max_heap'teki en büyük elemanın min_heap'teki en küçük elemandan küçük veya eşit olmasını sağlar.
        if self.min_heap and -self.max_heap[0] > self.min_heap[0]:
            val = -heapq.heappop(self.max_heap) # max_heap'ten en büyük elemanı al
            heapq.heappush(self.min_heap, val)  # min_heap'e ekle

        # Heap boyutlarını dengele:
        # max_heap, min_heap'e eşit veya bir eleman fazla olmalı.

        # Eğer max_heap çok büyükse (min_heap'ten birden fazla elemanı varsa), max_heap'ten min_heap'e taşı.
        if len(self.max_heap) > len(self.min_heap) + 1:
            val = -heapq.heappop(self.max_heap) # max_heap'ten en büyük elemanı al (negatif olduğu için en küçük indeksli)
            heapq.heappush(self.min_heap, val)  # min_heap'e ekle
        # Eğer min_heap çok büyükse (max_heap'ten fazla elemanı varsa), min_heap'ten max_heap'e taşı.
        elif len(self.min_heap) > len(self.max_heap):
            val = heapq.heappop(self.min_heap) # min_heap'ten en küçük elemanı al
            heapq.heappush(self.max_heap, -val) # max_heap'e negatifini ekle

    def findMedian(self) -> float:
        # Toplam sayı tek ise, medyan max_heap'in en büyük elemanıdır.
        # (max_heap her zaman eşit veya bir fazla elemana sahip olduğu için)
        if len(self.max_heap) > len(self.min_heap):
            return float(-self.max_heap[0]) # Negatif olarak saklandığı için tekrar negatife çevir
        # Toplam sayı çift ise, medyan iki heap'in en üstteki elemanlarının ortalamasıdır.
        else:
            if not self.max_heap: # Hiç sayı eklenmediyse
                return 0.0 # veya hata döndürülebilir
            return (-self.max_heap[0] + self.min_heap[0]) / 2.0


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
